cmd/utils/cmd.go

```diff
@@ -344,3 +346,141 @@ func ExportPreimages(db ethdb.Database, fn string) error {
 	log.Info("Exported preimages", "file", fn)
 	return nil
 }
+
+// ImportSnapshot imports a batch of snapshot data into the database
+func ImportChainData(db ethdb.Database, fn string, kind string, checker func(key []byte) bool, interrupt chan struct{}) error {
+	log.Info("Importing chain data", "file", fn, "kind", kind)
+
+	// Open the file handle and potentially unwrap the gzip stream
+	fh, err := os.Open(fn)
+	if err != nil {
+		return err
+	}
+	defer fh.Close()
+
+	var reader io.Reader = fh
+	if strings.HasSuffix(fn, ".gz") {
+		if reader, err = gzip.NewReader(reader); err != nil {
+			return err
+		}
+	}
+	stream := rlp.NewStream(reader, 0)
+
+	// Import the snapshot in batches to prevent disk trashing
+	var (
+		count    int64
+		filtered int64
+		start    = time.Now()
+		logged   = time.Now()
+		batch    = db.NewBatch()
+	)
+	for {
+		// Read the next entry and ensure it's not junk
+		var key, val []byte
+		if err := stream.Decode(&key); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if err := stream.Decode(&val); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if !checker(key) {
+			filtered += 1
+			continue
+		}
+		batch.Put(key, val)
+		if batch.ValueSize() > ethdb.IdealBatchSize {
+			if err := batch.Write(); err != nil {
+				return err
+			}
+			batch.Reset()
+		}
+		// Check interruption emitted by ctrl+c
+		if count%1000 == 0 {
+			select {
+			case <-interrupt:
+				if err := batch.Write(); err != nil {
+					return err
+				}
+				log.Info("Snapshot data importing interruptted", "file", fn, "kind", kind, "count", count, "filtered", filtered, "elapsed", common.PrettyDuration(time.Since(start)))
```

Comment:
Since we output a `count`, we could also have an optional `count` as input, so it can be resumed later.